### PR TITLE
Allows to load utxo set lazily, so OMG.State initializes quickly

### PR DESCRIPTION
**Pull request by pnowosie**
:clipboard: Related to #672, https://github.com/omisego/devops/issues/142

## Overview

This is continuation of @pdobacz's work in #1066 in regards to utxo lazy loading.

### How it works 
During `OMG.State` startup no UTXOs are fetched from DB, which is no longer blocking significantly.
Then during each of 6 utxo-related operations (see below) UTXO set is extended with UTXOs from DB to ensure operation behavior hasn't been changed.

Transaction processing is populating the in-memory UTXO set and once block is formed newly created UTXO are inserted to DB, but are also kept in process State. Service restart looses all UTXO created by transactions processed in the current block.

### Affected operations
Here I describe what changes was needed to maintain the previous behavior of the utxo-related functionality

* `utxo_exists?(utxo_pos)` - because `utxo_pos` in question could be created before process restarts its fetched from DB and merged to the in-memory UTXO set. On the other hand `utxo_pos` could be spend in recent transaction (in current, not-formed block) so it yet doesn't exists in-memory, but still exists in DB. Therefore new in-memory set `recently_spent` was added to handle this case.
* `exec(tx, _)` - It works the same as previously, but checks whether `tx`'s inputs exists is done in the DB-extended UTXO set.  It also populates `recently_spent` collection of the UTXOs spent in the current block.
* `deposit()` - no changes
* `exit_utxos(...)` - because `exit` operation supports multiple inputs forms (uto_pos, encoded utxo_pos, ife exits infos ...) it was split into two operations. First `Core.get_exiting_utxos` translates any of the inputs into exiting utxo position list. Then these UTXOs are fetched from DB and merged with in-memory set. Then exit from utxos are preformed. It also (just in case) populates the `recently_spent` collection of the exiting utxos.
* `form_block(...)` - works as previously, but also clears the `recently_spent` collection (because db_updates are pushed to DB)
* `close_block(...)` - no changes, it uses `form_block` so as mentioned :point_up:   

## Testing

New test are added to demonstrate behavior of the extended state. Fixes to existing tests were necessary.  

More information related to performance and how the change anticipate the future work can be find in description to :pr: #1066

### CODE REVIEW

File: apps/omg/lib/omg/state.ex [hunk 1]
```diff
@@ -90,21 +91,16 @@ defmodule OMG.State do
     # Get data essential for the State and Blockgetter. And it takes a while. TODO - measure it!
```
Comment (pdobacz) [inline apps/omg/lib/omg/state.ex]: update `@doc` perhaps? now it's too vague, let's make it specific.

Also `State.Core` module doc and `extract_initial_state` docs require update (check if anything else, incl. higher level docs, `architecture.md` maybe?). Maybe `extract_initial_state` should just be `init` instead

File: apps/omg/lib/omg/state.ex [hunk 2]
```diff
@@ -222,4 +224,17 @@ defmodule OMG.State do
     :ok = OMG.Bus.broadcast("events", {:preprocess_emit_events, event_triggers})
     :ok = OMG.Bus.direct_local_broadcast("blocks", {:enqueue_block, block})
   end
+
+  defp load_necessary_utxos_from_db(state, tx) do
+    tx
+    |> Transaction.get_inputs()
```
Comment (pdobacz) [inline apps/omg/lib/omg/state.ex]: I think everything except `Enum.map(&utxo_from_db/1)` should go to `core.ex` and have tests

File: apps/omg/lib/omg/state/core.ex [hunk 3]
```diff
@@ -99,31 +99,24 @@ defmodule OMG.State.Core do
...
-        _utxos_query_result,
-        :not_found,
-        _child_block_interval
-      ) do
+  def extract_initial_state(:not_found, _child_block_interval) do
     {:error, :top_block_number_not_found}
   end
 
+  @spec with_utxos(t(), UtxoSet.query_result_t()) :: t()
+  def with_utxos(%Core{utxos: utxos} = state, utxos_query_result),
```
Comment (pdobacz) [inline apps/omg/lib/omg/state/core.ex]: `defp`? also `exec` would have to be `defp`

OTOH let's contrast this approach with the more explicit `def with_utxos`, which enriches the UTXOset then `exec` which remains untouched. This has the benefit of smaller impact on the API & tests.

File: apps/omg/lib/omg/state/core.ex [hunk 4]
```diff
@@ -99,31 +99,24 @@ defmodule OMG.State.Core do
...
+    do: %{state | utxos: UtxoSet.merge_with_query_result(utxos, utxos_query_result)}
```
Comment (pdobacz) [inline apps/omg/lib/omg/state/core.ex]: ah, there's a more serious problem I missed in my experiments. If there are 2 txs that spend the same UTXO in a single block, they'll pass (there should be appropriate tests for this being impossible BTW, so the `State.CoreTest` should be expanded a lot). The first spends the UTXOs, but the second will pull them back from the DB again.

Looks like one needs to commit the `utxo_db_updates` to the DB on every `exec` in this case too.

Comment (pnowosie) [reply apps/omg/lib/omg/state/core.ex]: Ah, no - we couldn't commit the updates because if everything dies & restart before we form the block the state is corrupted.

Comment (pnowosie) [reply apps/omg/lib/omg/state/core.ex]: :bulb: Good idea: add test to show :point_up: problem.

File: apps/omg/lib/omg/state/core.ex (hunk 3 above)
Comment (pdobacz) [reply apps/omg/lib/omg/state/core.ex]: after discussion: public `with_utxos` and public `exec` seem more elegant. 

Just let's treat `exec` and other UTXOSet-operating functions assume that the UTXOSet operated on is well prepared. `with_utxos` prepares it, and must caare about all the edge cases (like the `recently_spent` - case discussed)

Let's try this

File: apps/omg/test/omg/state/persistence_test.exs [hunk 5]
```diff
@@ -198,7 +198,10 @@ defmodule OMG.State.PersistenceTest do
 
   defp persist_common(state, db_updates, db_pid) do
     assert :ok = OMG.DB.multi_update(db_updates, db_pid)
-    assert state == state_from(db_pid)
+    assert drop_recently_spent(state) == state_from(db_pid)
     state
   end
+
+  # NOTE: `recently_spent` is in-memory property of the %Core{}, which isn't persisted so it can't be recovered from db
+  defp drop_recently_spent(state), do: %Core{state | recently_spent: []}
```
Comment (pdobacz) [inline apps/omg/test/omg/state/persistence_test.exs]: eh, was going to comment to not reach into internals, but in fact this whole test is a bit silly right now. Let's stop and think what to do with it.

File: apps/omg_watcher/test/omg_watcher/exit_processor/core/state_interaction_test.exs [hunk 6]
```diff
@@ -194,9 +196,11 @@ defmodule OMG.Watcher.ExitProcessor.Core.StateInteractionTest do
     {:ok, %{^ife_id1 => exiting_positions1, ^ife_id2 => exiting_positions2}} =
       Core.prepare_utxo_exits_for_in_flight_exit_finalizations(processor, finalizations)
 
+    empty_db_utxos = %{}
```
Comment (pdobacz) [inline apps/omg_watcher/test/omg_watcher/exit_processor/core/state_interaction_test.exs]: again, the `with_utxos` argument here.

File: apps/omg/lib/omg/state/core.ex [hunk 7]
```diff
@@ -18,7 +18,7 @@ defmodule OMG.State.Core do
   All spend transactions, deposits and exits should sync on this for validity of moving funds.
   """
 
-  defstruct [:height, :utxos, pending_txs: [], tx_index: 0, utxo_db_updates: []]
+  defstruct [:height, utxos: %{}, pending_txs: [], tx_index: 0, utxo_db_updates: [], recently_spent: []]
```
Comment (pdobacz) [inline apps/omg/lib/omg/state/core.ex]: Let's
1/ revisit the docs to see if there aren't any obsolete stuffs there. This includes `moduledoc` here `docs` here `architecture.md`
2/ let's also expand the `moduledoc` here. I think now it requires a section explaning how the state is managed - what is in db, what in mem, what the semantics are there. Also - what happens if this crashes and restarts.

This part, let's tackle last, after the flows are nailed down

File: apps/omg/lib/omg/state.ex [hunk 8]
```diff
@@ -84,27 +86,14 @@ defmodule OMG.State do
   ### Server
 
   @doc """
-  Start processing state using the database entries
+  Initializes the state. UTXO set is not loaded now.
```
Comment (pdobacz) [inline apps/omg/lib/omg/state.ex]: don't have a better place to put the comment: we need to do sth about the utxo-based metrics calculations :(. Out of ideas for now

**Comment by InoMurko** [comment]
I'm trying to figure out what the purpose of `recently_spent` is. Can you help me with that? Wondering how we could simply this. 

Why are we loading UTXOs in the memory in the first place? Couldn't we maintain the list consistently in DB?

File: apps/omg/lib/omg/state.ex (hunk 8 above)
Comment (pnowosie) [reply apps/omg/lib/omg/state.ex]: I've created an isse for it #1148

**Comment by pnowosie** [comment]
> Why are we loading UTXOs in the memory in the first place? Couldn't we maintain the list consistently in DB and avoid loading them altogether?

It works like this:
Let's assume the chain has history of transactions and `OMG.State` process gets restarted for whatever reason.

During `OMG.State`  startup no UTXO will be loaded into memory - they all persist in DB.
Then process starts to process new transactions, deposits, exits... and it builds also it's in-memory utxoset.
These in-memory set works as an cache-layer as we don't need to fetch UTXO from DB if we know it was processed by State (`utxo position` exists either in `utxos` or `recently_spent`)

> I'm trying to figure out what the purpose of `recently_spent` is. Can you help me with that? 

So first, please mind that right now no info is written to DB after transaction is executed on `State`.
So let say `State` got restarted and it's processing two transactions spending the same UTXO.
As the input exists only in DB and DB isn't updated after tx is processed - one could be able to double spend - and that's why we need to store info about the recent spends. This collection gets cleared after block is formed because at that time all db updates are applied.

**Comment by InoMurko** [comment]
But why go through the difficulty of maintaining cache? Why do we need cache at all? I would say, ditch cache **completely**, fallback to strong consistency by using the DB on processing transactions. *The performance argument is irrelevant.*

File: apps/omg/lib/omg/state/core.ex [hunk 9]
```diff
@@ -250,77 +262,79 @@ defmodule OMG.State.Core do
...
-  Checks if utxo exists
+  Checks whether utxo exists in UTXO set.
+  Note: state passed here is already extended with DB.
   """
   @spec utxo_exists?(Utxo.Position.t(), t()) :: boolean()
-  def utxo_exists?(Utxo.position(_blknum, _txindex, _oindex) = utxo_pos, %Core{utxos: utxos}),
-    do: UtxoSet.exists?(utxos, utxo_pos)
+  def utxo_exists?(Utxo.position(_blknum, _txindex, _oindex) = utxo_pos, %Core{utxos: utxos}) do
+    utxos
+    |> UtxoSet.exists?(utxo_pos)
```
Comment (InoMurko) [inline apps/omg/lib/omg/state/core.ex]:
```suggestion
    |> UtxoSet.exists?(utxos, utxo_pos)
```

File: apps/omg/test/omg/state/core_test.exs [hunk 10]
```diff
@@ -49,6 +49,143 @@ defmodule OMG.State.CoreTest do
...
+    end
+
+    @tag fixtures: [:alice, :bob, :state_alice_deposit]
+    test "form block drops recent spends", %{alice: alice, bob: bob, state_alice_deposit: state} do
+      tx = create_recovered([{1, 0, 0, alice}], @eth, [{bob, 10}])
+      expected_spends = MapSet.new([Utxo.position(1, 0, 0)])
+
+      {:ok, _, state1} =
+        state
+        |> Core.exec(tx, :no_fees_required)
```
Comment (InoMurko) [inline apps/omg/test/omg/state/core_test.exs]:
```suggestion
        Core.exec(state, tx, :no_fees_required)
```

**Comment by pnowosie** [comment]
> I would say, ditch cache completely, fallback to strong consistency...

It would require to rewrite current behaviour completely which is risky. Currenlty all `State.Core` operations are done on processes state (I call it cache to draw an analogy) and therefore we can reuse a lot of existing tests. 

Also making all changes in DB requires to carefully consider all the cases, e.g. we would need to persist the mempool transactions...
What's more while this :pr: moves as closer to #849 it is here to address #672

File: apps/omg/lib/omg/state/core.ex (hunk 7 above)
Comment (pnowosie) [reply apps/omg/lib/omg/state/core.ex]: Extended @moduledoc - please see if it makes sense to you @pdobacz

**Comment by pdobacz** [comment]
> > I would say, ditch cache completely, fallback to strong consistency...
> 
> It would require to rewrite current behaviour completely which is risky.

Yeah, we confronted a couple of approaches here, ditching cache being one (and tempting!), but then it seemed like a much bigger change. So instead this is taking baby-steps instead, though I admit - not perfect.

File: apps/omg/test/omg/state/core_test.exs [hunk 11]
```diff
@@ -49,6 +49,132 @@ defmodule OMG.State.CoreTest do
...
+      |> success?()
+      # this is important - state is extended before applying tx
+      |> Core.with_utxos(db_utxos)
+      |> Core.exec(tx, :no_fees_required)
+      |> fail?(:utxo_not_found)
+    end
+
+    @tag fixtures: [:state_alice_deposit]
+    test "extending state with same utxos does not change it", %{state_alice_deposit: state} do
+      db_utxos = state.utxos
```
[the remark this answers is not in the document]
Comment (pdobacz) [reply apps/omg/test/omg/state/core_test.exs]: this

File: apps/omg/test/omg/state/core_test.exs [hunk 12]
```diff
@@ -49,6 +49,132 @@ defmodule OMG.State.CoreTest do
...
+
+      state
+      |> Core.with_utxos(db_utxos)
+      |> same?(state)
+    end
+
+    @tag fixtures: [:alice, :bob, :state_empty]
+    test "extending state partially", %{alice: alice, bob: bob, state_empty: state} do
+      fst_utxo = make_utxos([{1000, 0, 0, alice, @eth, 6}])
```
[the remark this answers is not in the document]
Comment (pdobacz) [reply apps/omg/test/omg/state/core_test.exs]: this, pls, look throught the comments from that review, seems like some of them got missed.

File: apps/omg/test/omg/state/core_test.exs [hunk 13]
```diff
@@ -49,6 +49,141 @@ defmodule OMG.State.CoreTest do
...
+    end
+
+    @tag fixtures: [:alice, :bob, :state_empty]
+    test "recent spends are stored regardless state extension", %{alice: alice, bob: bob, state_empty: state} do
+      init_state = do_deposit(state, alice, %{amount: 10, currency: @eth, blknum: 1})
+      tx1 = create_recovered([{1, 0, 0, alice}], @eth, [{bob, 10}])
+
+      {:ok, _, state1} = Core.exec(init_state, tx1, :no_fees_required)
+
+      %Core{recently_spent: spent1} = state1
```
Comment (pdobacz) [inline apps/omg/test/omg/state/core_test.exs]: now this is a bit awkward... why are we constantly adding cases where we test via internals? here and next `test`.

I'm guessing, you need to transform the tests into more behavior-driven - start with the name, the name of the test is coupled to internals. What is the behavior that we're after here?

**Comment by pnowosie** [comment]
> + pinging about the previous ones

Appologizes, GH is fooling me hidding comments especially on second review or is it rebase effects?

File: apps/omg/lib/omg/state.ex [hunk 14]
```diff
@@ -222,4 +224,17 @@ defmodule OMG.State do
...
+  @spec init_utxos_from_db(list(InputPointer.Protocol.t())) :: UtxoSet.t()
+  defp init_utxos_from_db(utxo_pos_list) do
+    utxo_pos_list
+    |> Enum.map(&utxo_from_db/1)
+    |> Enum.reject(&(:not_found == &1))
+    |> UtxoSet.init()
+  end
+
+  defp utxo_from_db(input_pointer) do
+    with {:ok, utxo_kv} <- DB.utxo(InputPointer.Protocol.to_db_key(input_pointer)),
```
[the remark this answers is not in the document]
Comment (pnowosie) [reply apps/omg/lib/omg/state.ex]: As we talked - we need to handle :not_found. Now they're filtered out in UtxoSet.init

File: apps/omg/test/omg/state/core_test.exs (hunk 13 above)
Comment (pnowosie) [reply apps/omg/test/omg/state/core_test.exs]: I drop these tests as this is an detail but we actually don't have to clear `recently_spent` on form_block

**Comment by pnowosie** [comment]
@InoMurko FYI
> But why go through the difficulty of maintaining cache?

I created an :issue: #1156 that proposed next simple* steps to ditch the cache completely. 
*) it looks simple on paper :smile:

File: apps/omg/lib/omg/state/core.ex [hunk 15]
```diff
@@ -16,9 +16,38 @@ defmodule OMG.State.Core do
...
+  on process startup.
+
+  During OMG.State startup no UTXOs are fetched from DB, which is no longer blocking significantly.
+  Then during each of 6 utxo-related operations (see below) UTXO set is extended with UTXOs from DB to ensure operation
+  behavior hasn't been changed.
+
+  Transaction processing is populating the in-memory UTXO set and once block is formed newly created UTXO are inserted
+  to DB, but are also kept in process State. Service restart looses all UTXO created by transactions processed as well
+  as mempool transactions therefore DB content stays block-by-block consistent.
+  the current block.
```
Comment (pdobacz) [inline apps/omg/lib/omg/state/core.ex]: leftover?

File: apps/omg/lib/omg/state/core.ex [hunk 16]
```diff
@@ -41,7 +70,10 @@ defmodule OMG.State.Core do
           tx_index: non_neg_integer(),
           # NOTE: that this list is being build reverse, in some cases it may matter. It is reversed just before
           #       it leaves this module in `form_block/3`
-          utxo_db_updates: list(db_update())
+          utxo_db_updates: list(db_update()),
+          # NOTE: because UTXO set is not loaded from DB entirely, we need to remember the UTXOs spent in already
+          # processed transaction before they get removed from DB on form_block.
+          recently_spent: MapSet.t(Utxo.Position.t())
```
Comment (pdobacz) [inline apps/omg/lib/omg/state/core.ex]: shouldn't be `(InputPointer.Protocol.t)` here?

File: apps/omg/lib/omg/state/core.ex [hunk 17]
```diff
@@ -96,34 +128,41 @@ defmodule OMG.State.Core do
...
-        _utxos_query_result,
-        :not_found,
-        _child_block_interval
-      ) do
+  def extract_initial_state(:not_found, _child_block_interval) do
     {:error, :top_block_number_not_found}
   end
 
+  @doc """
+  Tell whether utxo position was created or spend by current state.
```
Comment (pdobacz) [inline apps/omg/lib/omg/state/core.ex]: `spent`

File: apps/omg/lib/omg/state/core.ex [hunk 18]
```diff
@@ -96,34 +128,41 @@ defmodule OMG.State.Core do
...
+  It's an optimization as there is no need to ask DB of known utxos
```
Comment (pdobacz) [inline apps/omg/lib/omg/state/core.ex]: drop this line, it is not an optimization, if you don't `or` with `recently_spent` you'll allow double spends.

File: apps/omg/test/omg/state/persistence_test.exs [hunk 19]
```diff
@@ -33,172 +34,205 @@ defmodule OMG.State.PersistenceTest do
...
+    {:ok, _, _} = OMG.State.exit_utxos([Utxo.position(1, 0, 0)])
 
-    state
+    # exit db_updates won't get persisted yet, but alice tries to spent it immediately
+    assert :utxo_not_found == exec(create_recovered([{1, 0, 0, alice}], @eth, [{alice, 20}]))
+
+    # retry above with empty in-memory utxoset
+    :ok = restart_state()
+
+    {:ok, _, _} = OMG.State.exit_utxos([Utxo.position(1, 0, 0)])
```
Comment (pdobacz) [inline apps/omg/test/omg/state/persistence_test.exs]: hm, this shouldn't be necessary and could harm the test, if you weren't persisting

File: apps/omg/test/omg/state/persistence_test.exs [hunk 20]
```diff
@@ -33,172 +34,205 @@ defmodule OMG.State.PersistenceTest do
...
+    # easiest to do this is to schedule another operation on `OMG.State` which is blocking
+    _ = OMG.State.utxo_exists?(Utxo.position(0, 0, 0))
+
+    :ok
   end
 
-  defp persist_exit_utxos(state, utxo_positions, db_pid) do
-    assert {:ok, {db_updates, _}, state} = utxo_positions |> Core.exit_utxos(state)
-    persist_common(state, db_updates, db_pid)
+  defp exec(:ok, tx), do: exec(tx)
```
Comment (pdobacz) [inline apps/omg/test/omg/state/persistence_test.exs]: oh, wait, in the previous `persistence_test` we were piping to transfer `state`, now we don't have anything to pipe through the sequence of steps. So you ended up extending the piped functions to handle the `:ok`'s, but you could just unpipe the whole thing :joy:. I feel a bit culpable for this old test, so if you like, I can take care of it.

oh, so I also had the thought to merge these tests into `state_test.exs` and tidy that up too, so that there is something resembling a proper test of `OMG.State`. But you can punt this on me too, just LMK

Comment (pnowosie) [reply apps/omg/test/omg/state/persistence_test.exs]: At first I unpiped the flow, but then I saw the pipe approach as  `assert the previous  operation succeeded` because we don't have the `:error` conterpart

File: apps/omg/lib/omg/state/core.ex (hunk 16 above)
Comment (pnowosie) [reply apps/omg/lib/omg/state/core.ex]: :+1:

File: apps/omg/lib/omg/state.ex [hunk 21]
```diff
@@ -132,7 +122,15 @@ defmodule OMG.State do
...
   """
   def handle_call({:exec, tx, fees}, _from, state) do
-    case Core.exec(state, tx, fees) do
+    db_utxos =
+      tx
+      |> Transaction.get_inputs()
+      |> init_utxos_from_db(state)
+
+    state
+    |> Core.with_utxos(db_utxos)
```
Comment (InoMurko) [inline apps/omg/lib/omg/state.ex]: this seems to be a bit confusing step. why is this step separate from processing a transaction?

File: apps/omg/test/omg/state/persistence_test.exs (hunk 19 above)
Comment (pnowosie) [reply apps/omg/test/omg/state/persistence_test.exs]: IMHO the test is valid as it's race condition possible *persisting exit is done outside of  `OMG.State` process) 
Let say `OMG.State` will have two requests (messages in its inbox)
1. :exit_utxo([alice_utxo])
2. :exec(tx_from_alice_spending_utxo)

So the test shows that even on this edge case there will be no double spend
WDYT?

File: apps/omg/lib/omg/state.ex [hunk 22]
```diff
@@ -156,7 +154,12 @@ defmodule OMG.State do
   Exits (spends) utxos on child chain, explicitly signals all utxos that have already been spent
   """
   def handle_call({:exit_utxos, utxos}, _from, state) do
-    {:ok, {db_updates, validities}, new_state} = Core.exit_utxos(utxos, state)
+    exiting_utxos = Core.get_exiting_utxo_positions(utxos, state)
+
+    db_utxos = init_utxos_from_db(exiting_utxos, state)
```
Comment (InoMurko) [inline apps/omg/lib/omg/state.ex]: `init` prefix seems to be used a couple of times... that invalidates the meaning.

File: apps/omg/lib/omg/state/core.ex [hunk 23]
```diff
@@ -250,75 +291,89 @@ defmodule OMG.State.Core do
   NOTE: It is done like this to accommodate different clients of this function as they can either be
   bare `EthereumEventListener` or `ExitProcessor`. Hence different forms it can get the exiting utxos delivered
   """
-  @spec exit_utxos(exiting_utxos :: exiting_utxos_t(), state :: t()) ::
-          {:ok, {[db_update], validities_t()}, new_state :: t()}
-  # empty list of whatever to bypass typing
-  def exit_utxos([], %Core{} = state), do: {:ok, {[], {[], []}}, state}
+  @spec get_exiting_utxo_positions(exiting_utxos_t(), t()) :: list(Utxo.Position.t())
+  def get_exiting_utxo_positions(exit_infos, state)
```
Comment (InoMurko) [inline apps/omg/lib/omg/state/core.ex]: the correct naming (since there's no IO) would be *extract*

File: apps/omg/test/omg/state/persistence_test.exs (hunk 19 above)
Comment (pdobacz) [reply apps/omg/test/omg/state/persistence_test.exs]: oh right, you're not persisting here, gotcha

File: apps/omg/lib/omg/state.ex (hunk 14 above)
[the remark this answers is not in the document]
Comment (InoMurko) [reply apps/omg/lib/omg/state.ex]: maybe a better way of handling this would be to let the developer know what output's you're able to handle
```
case DB.utxo(InputPointer.Protocol.to_db_key(input_pointer)) do
  :not_found -> :not_found
  {:ok, _utxo_kv} = result -> result
end
```
this means that if you get a value back you're not able to handle, you would *fail and crash early*

File: apps/omg/lib/omg/state/core.ex [hunk 24]
```diff
@@ -16,9 +16,37 @@ defmodule OMG.State.Core do
...
+
+  We experienced long startup times on large UTXO set, which in some case caused timeouts and lethal `OMG.State`
+  restart loop. To mitigate this issue we introduced loading UTXO set on demand (see GH#1103) instead of full load
+  on process startup.
+
+  During OMG.State startup no UTXOs are fetched from DB, which is no longer blocking significantly.
+  Then during each of 6 utxo-related operations (see below) UTXO set is extended with UTXOs from DB to ensure operation
+  behavior hasn't been changed.
+
+  Transaction processing is populating the in-memory UTXO set and once block is formed newly created UTXO are inserted
```
Comment (InoMurko) [inline apps/omg/lib/omg/state/core.ex]: side question:
```
Transaction processing is populating the in-memory UTXO set and once block is formed newly created UTXO are inserted to DB, but are also kept in process State.
```
what happens if an outage happens in between? Between block is formed - newly created UTXOs are inserted to DB?
1. What if you're not able to submit a block, but were able to persist to DB.
2. What if you're able to submit a block, but not able to persist to DB.

File: apps/omg/lib/omg/state/core.ex [hunk 25]
```diff
@@ -250,75 +289,89 @@ defmodule OMG.State.Core do
...
     |> Enum.flat_map(fn %{call_data: %{in_flight_tx: tx_bytes}} ->
       {:ok, tx} = Transaction.decode(tx_bytes)
       Transaction.get_inputs(tx)
     end)
-    |> exit_utxos(state)
   end
 
   # list of IFE input piggybacked events (they're ignored)
-  def exit_utxos([%{tx_hash: _, omg_data: %{piggyback_type: :input}} | _] = piggybacks, state) do
+  def get_exiting_utxo_positions([%{tx_hash: _, omg_data: %{piggyback_type: :input}} | _] = piggybacks, %Core{}) do
```
Comment (InoMurko) [inline apps/omg/lib/omg/state/core.ex]: useless matching on `%Core{}` ?

File: apps/omg/lib/omg/state/core.ex [hunk 26]
```diff
@@ -250,75 +289,89 @@ defmodule OMG.State.Core do
...
+  """
+  @spec exit_utxos(exiting_utxos :: list(Utxo.Position.t()), state :: t()) ::
+          {:ok, {[db_update], validities_t()}, new_state :: t()}
+  def exit_utxos([], %Core{} = state), do: {:ok, {[], {[], []}}, state}
+
+  def exit_utxos(
+        [Utxo.position(_, _, _) | _] = exiting_utxos,
+        %Core{utxos: utxos, recently_spent: recently_spent} = state
+      ) do
     _ = Logger.info("Recognized exits #{inspect(exiting_utxos)}")
```
Comment (InoMurko) [inline apps/omg/lib/omg/state/core.ex]: Dangerous logging if large number of exits!

File: apps/omg/lib/omg/state/core.ex [hunk 27]
```diff
@@ -250,75 +289,89 @@ defmodule OMG.State.Core do
...
-    |> exit_utxos(state)
   end
 
   # list of utxo positions (decoded)
-  def exit_utxos([Utxo.position(_, _, _) | _] = exiting_utxos, %Core{utxos: utxos} = state) do
+  def get_exiting_utxo_positions([Utxo.position(_, _, _) | _] = exiting_utxos, %Core{}), do: exiting_utxos
+
+  @doc """
+  Spends exited utxos.
+  Note: state passed here is already extended with DB.
```
Comment (InoMurko) [inline apps/omg/lib/omg/state/core.ex]: the reason why you had to leave this comment is because you introduced `Core.with_utxos/2` which makes it harder to reason about what it does with state!

File: apps/omg/test/omg/state/persistence_test.exs [hunk 28]
```diff
@@ -33,172 +34,205 @@ defmodule OMG.State.PersistenceTest do
...
-    {:ok, {_, _, db_updates}, state} = Core.form_block(@interval, state)
-    persist_common(state, db_updates, db_pid)
+  defp persist_form(:ok), do: persist_form()
+
+  defp persist_form() do
+    OMG.State.form_block()
+
+    # because `form_block` is non-blocking operation we need to wait it finishes
+    # easiest to do this is to schedule another operation on `OMG.State` which is blocking
+    _ = OMG.State.utxo_exists?(Utxo.position(0, 0, 0))
```
Comment (InoMurko) [inline apps/omg/test/omg/state/persistence_test.exs]: I don't understand this. Can you please elaborate what `persist_form` does?

File: apps/omg/test/omg/state/persistence_test.exs [hunk 29]
```diff
@@ -33,172 +34,205 @@ defmodule OMG.State.PersistenceTest do
...
+
+    :ok = OMG.DB.multi_update(db_updates)
   end
 
-  defp persisted_standard_exitable_utxos(address, db_pid) do
-    {:ok, utxos_query_result} = OMG.DB.utxos(db_pid)
-    Core.standard_exitable_utxos(utxos_query_result, address.addr)
+  defp make_deposits(list) do
+    list
+    |> Enum.map(fn %{owner: owner, currency: currency, amount: amount, blknum: blknum} ->
```
Comment (InoMurko) [inline apps/omg/test/omg/state/persistence_test.exs]:
```suggestion
    Enum.map(list, fn %{owner: owner, currency: currency, amount: amount, blknum: blknum} ->
```

File: apps/omg/test/omg/state/persistence_test.exs [hunk 30]
```diff
@@ -33,172 +34,205 @@ defmodule OMG.State.PersistenceTest do
...
+    end)
   end
 
-  defp persist_common(state, db_updates, db_pid) do
-    assert :ok = OMG.DB.multi_update(db_updates, db_pid)
-    assert state == state_from(db_pid)
-    state
+  defp restart_state() do
+    GenServer.stop(OMG.State)
+    WaitFor.ok(fn -> if(GenServer.whereis(OMG.State), do: :ok) end)
```
Comment (InoMurko) [inline apps/omg/test/omg/state/persistence_test.exs]: `stop/1,2` is a synchronous operation.
https://hexdocs.pm/elixir/GenServer.html#stop/3

File: apps/omg/test/omg/state/utxo_set_test.exs [hunk 31]
```diff
@@ -63,6 +63,16 @@ defmodule OMG.State.UtxoSetTest do
                |> UtxoSet.init()
                |> UtxoSet.get_by_inputs(inputs)
     end
+
+    test "ignores db query possible not_founds",
```
Comment (InoMurko) [inline apps/omg/test/omg/state/utxo_set_test.exs]: I'm having difficulties understanding the purpose of this test... a better description would help.

File: apps/omg/test/omg/state_test.exs [hunk 32]
```diff
@@ -51,7 +51,8 @@ defmodule OMG.StateTest do
   @tag fixtures: [:alice, :standalone_state_server]
   test "can execute various calls on OMG.State, one happy path only", %{alice: alice} do
```
Comment (InoMurko) [inline apps/omg/test/omg/state_test.exs]: I'm having difficulties understanding the purpose of this test... can't see why your changes would affect the test.

File: docs/architecture.md [hunk 33]
```diff
@@ -103,7 +103,7 @@ Each instance of either Child Chain Server or Watcher should have it's own insta
 
 Database necessary to properly ensure validity and availability of blocks and transactions
 
-- it is read by `OMG.State` to discover the UTXO set on restart
+- it is read by `OMG.State`  on restart to discover the current block number and Ethereum height, UTXO set is lazily loaded
```
Comment (InoMurko) [inline docs/architecture.md]: this could be formulated better. Also... this document is pretty stale. might be better if we retire this one until we have something more firm.

File: apps/omg/test/omg/state/persistence_test.exs (hunk 30 above)
Comment (pnowosie) [reply apps/omg/test/omg/state/persistence_test.exs]: That's correct - `WaitFor` is here to assume the supervisor bring the process back

Comment (InoMurko) [reply apps/omg/test/omg/state/persistence_test.exs]: Why the assert? 
`assert OMG.State |> GenServer.whereis() |> Kernel.is_pid()`

Comment (pnowosie) [reply apps/omg/test/omg/state/persistence_test.exs]: To double check :smile: 
I'll drop the `assert` line

File: apps/omg/test/omg/state_test.exs (hunk 32 above)
Comment (pnowosie) [reply apps/omg/test/omg/state_test.exs]: Ouch left off from initial branch - reverted.
Thx!

File: apps/omg/lib/omg/state.ex (hunk 21 above)
Comment (pnowosie) [reply apps/omg/lib/omg/state.ex]: Well extending the state of the utxos from DB is in several places, not just for tx processing.

File: apps/omg/lib/omg/state/core.ex (hunk 24 above)
Comment (pnowosie) [reply apps/omg/lib/omg/state/core.ex]: Case 2. is impossible. I'll describe case 1:

There are 2 processes engaged to block publishing
* OMG.State - check `form_block`
* OMG.ChildChain.BlockQueue - it check what gets mined on Ethereum and retries if needed

All _artifacts_ (mempool txs, new utxo, spends, block) from the OMG.State form_block is inserted to DB in single transaction (`multi_update`).
Then it publishes the event that new block is available - BQ is listening on these events.
If the outage blocks the event :point_up: propagation, BQ on init will check what gets mined on Ethereum and what block we have in DB and resend the missings.

Am I answered your concern?
